drivers: at_cmd: optimization and bug fixes

evenl opened the pull request:
This PR contains multiple fixes discovered after extensive testing of the at_cmd driver to hunt down an issue. I have organized the fixes in separated commits.

Summary:

* Switch from mutex to semaphore to protect critical sections
* Don't submit a worker if the callback is NULL
* Don't modify ret.state if error detected is not AT related
* Only fail on "true" error and return the errno

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=958) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicPlayground/fw-nrfconnect-nrf?pullRequest=958) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicPlayground/fw-nrfconnect-nrf?pullRequest=958) it.</sub>

NordicBuilder commented:
Found the following issues, please fix and resubmit:

## Identity/Emails issues

6733c0e0195f1398f4742a286ca826a6d648ced6: author email (Even Falch-Larsen <even.falch-larsen@nordicsemi.no>) needs to match one of the signed-off-by entries.

0269a2ad82faf4aa8cbdaeca84f0f1c66c878086: author email (Even Falch-Larsen <even.falch-larsen@nordicsemi.no>) needs to match one of the signed-off-by entries.

3c383292a374da4b7b344ecdb1436218607a81ae: author email (Even Falch-Larsen <even.falch-larsen@nordicsemi.no>) needs to match one of the signed-off-by entries.

d97c42091be0a1b28778b30b43a7c55b3ec907d7: author email (Even Falch-Larsen <even.falch-larsen@nordicsemi.no>) needs to match one of the signed-off-by entries.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -284,27 +301,34 @@ int at_cmd_write(const char *const cmd,
 		 enum at_cmd_state *state)
 {
-	int return_code;
+	int return_code = k_sem_take(&cmd_pending, K_FOREVER);
```
`k_sem_take` will not return an error if it's called with `K_FOREVER` timeout. It can only return an error code, in case it times-out (-EAGAIN), or when it was called with `K_NO_WAIT` (-EBUSY). Hence, I'd remove error checks to simplify the logic.

evenl replied:
Ok, removed the code in this function, at_cmd_write_with_callback() and at_cmd_set_notification_handler

evenl commented:
I completely agree with you that the mutex semantics fits this use-case much better than a semaphore, and I was kind of reluctant to switch over to semaphores. The reason for doing that is that there is no API for checking if someone is waiting for the muxtex to be unlocked, while on the semaphore API you can do it indirectly (when using it as binary semaphore) by looking at the count. I changed line 149/150, from a hack to using proper API calls. So for maintainablity i changed it to using semaphore instead.

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -267,14 +281,13 @@ int at_cmd_write_with_callback(const char *const cmd,
 			       at_cmd_handler_t  handler,
 			       enum at_cmd_state *state)
 {
-	int return_code;
```
You need to keep the declaration.

evenl replied:
Yes, I was a bit quick there :D

rlubos commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -284,27 +297,26 @@ int at_cmd_write(const char *const cmd,
 		 size_t buf_len,
 		 enum at_cmd_state *state)
 {
-	int return_code;
```
Same here.

rlubos commented:
@evenl I've allowed myself to rebase your branch, it should pass the CI now...

metc commented on drivers/at_cmd/at_cmd.c:
```diff
@@ -110,7 +111,7 @@ static void callback_worker(struct k_work *item)
 	struct callback_work_item *data =
 		CONTAINER_OF(item, struct callback_work_item, work);
 
-	if ((data != NULL) && (data->callback != NULL)) {
```
I think it is safer to keep this check.

rlubos replied:
`data` is still checked, and `data->callback` is not supposed to be NULL after these changes, so we're safe here. We could optionally assert on `data->callback`, but I think we're good to go for now.
